js/modules/k6/experimental/fs/errors.go

```diff
@@ -0,0 +1,58 @@
+package fs
+
+// newFsError creates a new Error object of the provided kind and with the
+// provided message.
+func newFsError(k errorKind, message string) *fsError {
+	return &fsError{
+		Name:    k.String(),
+		Message: message,
+		kind:    k,
+	}
+}
+
+// errorKind indicates the kind of file system error that has occurred.
+//
+// Its string representation is generated by the `enumer` tool. The
+// `enumer` tool is run by the `go generate` command. See the `go generate`
+// command documentation.
+// The tool itself is not tracked as part of the k6 go.mod file, and
+// therefore must be installed manually using `go install github.com/dmarkham/enumer`.
+//
+//go:generate enumer -type=errorKind -output errors_gen.go
+type errorKind uint8
+
+const (
+	// NotFoundError is emitted when a file is not found.
+	NotFoundError errorKind = iota + 1
+
+	// InvalidResourceError is emitted when a resource is invalid: for
+	// instance when attempting to open a directory, which is not supported.
+	InvalidResourceError
+
+	// ForbiddenError is emitted when an operation is forbidden.
+	ForbiddenError
+
+	// TypeError is emitted when an incorrect type has been used.
+	TypeError
+)
+
+// fsError represents a custom error object emitted by the fs module.
+type fsError struct {
```

Comment:
The `fsError` type exists in order to return `Error`-like Javascript objects on failure with a predefined structure (in this case, an error name and a message). That also makes testing easier as I know exactly what an error would look like when I catch one.

I did this in WebCrypto to have clear error types with a predefined structure too. I had looked into alternatives to this way of doing and hadn't found a satisfying one, but if you have an alternative recommendation, I'm happy to give it a shot 👍🏻 

Deno, which this API is inspired from, also has a code property on the error objects returned by their File API, which might be something to add too.

I'll add tests asserting the errors are of the right type with `errors.As` indeed 🙇🏻

Reply:
>The fsError type exists in order to return Error-like Javascript objects on failure with a predefined structure

Are you aware if there is some sort of Deno-like or modern-JS defined somewhere about it that I could read?

>That also makes testing easier as I know exactly what an error would look like when I catch one.

I see you are doing
```
if (err.name !== 'TypeError')
```
Is this idiomatic JavaScript? I see [Deno is using `instanceof`](https://deno.land/api@v1.36.1?s=Deno.errors) for doing something similar. Maybe we should do the same.

Reply:
>I did this in WebCrypto to have clear error types with a predefined structure too.

Reading and recalling it, it could be the right moment for creating our opinionated struct and doing something generic in `js` package and reusing it across the modules instead of redefining it.

Reply:
> Is this idiomatic JavaScript? I see [Deno uses instanceof](https://deno.land/api@v1.36.1?s=Deno.errors) for doing something similar. Maybe we should do the same.

My understanding is that the alternative is tricky in Goja (at least from the last time I attempted it). Supporting `isinstanceof` implies setting the prototype of an Object to match the right-hand side of the expression. I looked into it in the context of a previous pull request. I remember we discussed it with @olegbespalov. Had looked into it at the time but couldn't find the way to do that in Goja during the time I allocated to it (it's probably doable, but I struggled to find a way to do it with Goja lacking examples on that front).

I'll give it another shot 🤞🏻 

> creating our opinionated struct, doing something generic in the js package, and reusing it across the modules instead of redefining it.

I've been considering doing it for a while now, but I don't have a clear path forward yet. I'll open an issue 👍🏻

Reply:
I gave it another shot, and in the two hours I spent on it, I came somewhat close, but I couldn't get `e instanceof FSError` to work 😞 

The main reason is that `instanceof` expects its right-hand side to be a function, and at the moment, Goja does not expose the ability to instantiate new native functions. If you check Goja's code, with the implementation of the `TypeError`, for instance, which is natively supported by Goja, you'll see that it depends on a lot of APIs that are not publicly exposed.

I have tried different approaches to try to work around that limitation, but I couldn't get any to work in the end:
- one approach could be to replicate the code that Goja uses internally in k6 (through my quick look, I didn't spot any blockers on that path), but it isn't necessarily a good idea.
- another approach would be to make a JS expression directly executed in the runtime.

I find both approaches unsatisfactory, and they would require some work that's out of the scope of this PR. 

I will open an issue to track this problematic, and keep the current implementation as-is.

Let me know if that works for you @codebien 🙇🏻

Reply:
I have opened an [issue with goja](https://github.com/dop251/goja/issues/529) 👍🏻 

Until there is a sensible solution to this, I will keep the implementation as is.

Resolving this conversation, feel free to re-open it if something changes 🙇🏻